File: frontend/src/types/project.ts
/**
 * The project lifecycle contract.
 *
 * Mirrors ``src/content_factory/models/project.py``. Three facts the studio is
 * written against, each of which a hand-written guess got wrong before:
 *
 * * ``Project.script`` is the **raw narration text** (a string) — it is what the
 *   renderer narrates and the linter analyses. The bundle an editor binds to
 *   (topic, style, parsed sections, timing plan) is ``Project.script_document``.
 * * ``Project.video`` is a :interface:`VideoAsset`, not a URL string.
 * * the publish payload is ``{platforms: [...]}``, not ``{destinations: [...]}``.
 *
 * Types that belong to another domain module are *not* duplicated here: scenes
 * live in ``./timeline``, the script bundle in ``./script``, research in
 * ``./research``. That keeps each concept in exactly one place and keeps the
 * barrel in ``./index`` free of colliding export names.
 */

import { ApprovalRecord, ProjectStatus } from "./common";
import { ScriptAnalysis, ScriptDocument, ScriptPlan, ScriptIssue } from "./script";
import { ResearchBundle } from "./research";
import { VideoProject } from "./timeline";
import { VoiceoverBundle } from "./voice";
import { Workflow } from "./workflow";
import { MultiFormatCampaign } from "./campaign";
import { ExternalAssetRecord } from "./external";
import {
  FactReconciliationReport,
  SensitivityAuditReport,
  StructuredTimeline,
} from "./history";

/** Payload for ``POST /projects`` (`ProjectCreate`). */
export interface ProjectCreate {
  name: string;
  topic: string;
  target_language?: string;
  duration_target_seconds?: number;
}

/**
 * Payload for ``PUT /projects/{id}/script`` (`ScriptUpdate`).
 *
 * ``script`` is the canonical field and ``raw_script`` is the alias the studio
 * sends; the backend accepts either, and the model keeps them in sync.
 */
export interface ScriptUpdate {
  script?: string | null;
  raw_script?: string | null;
  source_rights_confirmed?: boolean;
}

/** Payload for ``POST /projects/{id}/approvals`` (`ApprovalCreate`). */
export interface ApprovalCreate {
  stage: "script" | "video";
  verdict: "approved" | "rejected";
  comment?: string | null;
}

/** Payload for ``POST /projects/{id}/publish`` (`PublishCreate`). */
export interface PublishCreate {
  /** Destinations to publish to. Omitted means the default (``youtube``). */
  platforms?: string[];
}

/** The produced video asset (`VideoAsset`). */
export interface VideoAsset {
  asset_url: string;
  thumbnail_url: string;
  duration_seconds: number;
  format: string;
  size_bytes?: number | null;
}

/** A document attached to a project (`DocumentRef`). */
export interface DocumentRef {
  id: string;
  title: string;
  source: string;
  doi?: string | null;
  url?: string | null;
  pdf_path?: string | null;
  added_at: string;
}

/**
 * A project (`Project`).
 *
 * Every field the backend serialises is listed; a given screen reads a subset.
 * Fields that are nullable or defaulted on the backend stay optional here, so no
 * screen assumes a value the API did not send.
 */
export interface Project {
  id: string;
  name: string;
  topic: string;
  target_language: string;
  duration_target_seconds: number;
  status: ProjectStatus;
  /** Raw narration text — not the structured bundle. */
  script?: string | null;
  /** The structured view of ``script``, for editors. */
  script_document?: ScriptDocument | null;
  source_rights_confirmed: boolean;
  approvals: ApprovalRecord[];
  provider_used?: string | null;
  error?: string | null;
  progress?: number | null;
  video?: VideoAsset | null;
  platforms: string[];
  published_at?: string | null;
  research?: ResearchBundle | null;
  documents: DocumentRef[];
  knowledge_base_id?: string | null;
  video_project?: VideoProject | null;
  voiceover?: VoiceoverBundle | null;
  script_style: string;
  script_plan?: ScriptPlan | null;
  script_issues: ScriptIssue[];
  analysis?: ScriptAnalysis | null;
  agent_used?: string | null;
  workflow?: Workflow | null;
  campaign?: MultiFormatCampaign | null;
  external_assets: ExternalAssetRecord[];
  structured_timeline?: StructuredTimeline | null;
  fact_report?: FactReconciliationReport | null;
  sensitivity_report?: SensitivityAuditReport | null;
  source_media_id?: string | null;
  created_at: string;
  updated_at: string;
}

File: src/content_factory/models/project.py
"""Project aggregate and its lifecycle request payloads."""

from __future__ import annotations

from datetime import datetime

from pydantic import BaseModel, Field, model_validator

from .campaign import MultiFormatCampaign
from .common import (
    ApprovalRecord,
    ApprovalStage,
    ApprovalVerdict,
    ProjectStatus,
    utcnow,
)
from .external import ExternalAssetRecord
from .history import (
    FactReconciliationReport,
    SensitivityAuditReport,
    StructuredTimeline,
)
from .knowledge import GroundingBundle
from .research import (
    DocumentRef,
    ResearchBundle,
)
from .script import (
    ScriptDocument,
    ScriptIssue,
    ScriptPlan,
)
from .timeline import VideoProject
from .voice import VoiceoverBundle
from .workflow import Workflow


class ProjectCreate(BaseModel):
    """Payload for creating a project."""

    name: str = Field(min_length=1, max_length=120)
    topic: str = Field(min_length=1, max_length=500)
    target_language: str = "vi"
    duration_target_seconds: int = Field(default=45, ge=5, le=600)


class ScriptUpdate(BaseModel):
    """Payload for saving a script and confirming source rights.

    ``script`` is the canonical field; ``raw_script`` is the name both web clients
    send, because they think of it as the raw text behind the parsed sections.
    Exactly one of the two is required.
    """

    script: str | None = None
    raw_script: str | None = None
    source_rights_confirmed: bool = False

    @property
    def text(self) -> str:
        """The script body, whichever field carried it."""
        return self.script if self.script is not None else (self.raw_script or "")

    @model_validator(mode="after")
    def _require_a_script(self) -> ScriptUpdate:
        if not self.text.strip():
            raise ValueError("provide 'script' (or its alias 'raw_script')")
        return self


class ApprovalCreate(BaseModel):
    """Payload for submitting a human review decision."""

    stage: ApprovalStage
    verdict: ApprovalVerdict
    comment: str | None = None


class VideoAsset(BaseModel):
    """Metadata for a produced video."""

    asset_url: str
    thumbnail_url: str
    duration_seconds: int
    format: str = "mp4"
    size_bytes: int | None = None


class PublishCreate(BaseModel):
    """Payload for publishing an approved video."""

    platforms: list[str] = Field(default_factory=lambda: ["youtube"])


class Project(BaseModel):
    """A content production unit tracked through the pipeline."""

    id: str
    name: str
    topic: str
    target_language: str
    duration_target_seconds: int
    status: ProjectStatus = ProjectStatus.DRAFT
    script: str | None = None
    #: Structured view of ``script`` (sections, timing, style) for editors.
    #: ``script`` itself stays the raw text every other consumer expects.
    script_document: ScriptDocument | None = None
    source_rights_confirmed: bool = False
    approvals: list[ApprovalRecord] = Field(default_factory=list)
    provider_used: str | None = None
    error: str | None = None
    progress: int | None = None
    video: VideoAsset | None = None
    platforms: list[str] = Field(default_factory=list)
    published_at: datetime | None = None
    research: ResearchBundle | None = None
    documents: list[DocumentRef] = Field(default_factory=list)
    # RAGFlow-style knowledge base attached to this project.
    knowledge_base_id: str | None = None
    # Last grounding pass: retrieved chunks that grounded the current script.
    grounding: GroundingBundle | None = None
    video_project: VideoProject | None = None
    voiceover: VoiceoverBundle | None = None
    script_style: str = "viral-short"
    script_plan: ScriptPlan | None = None
    script_issues: list[ScriptIssue] = Field(default_factory=list)
    agent_used: str | None = None
    workflow: Workflow | None = None
    campaign: MultiFormatCampaign | None = None
    external_assets: list[ExternalAssetRecord] = Field(default_factory=list)
    structured_timeline: StructuredTimeline | None = None
    fact_report: FactReconciliationReport | None = None
    sensitivity_report: SensitivityAuditReport | None = None
    # Media-library item this project was re-cooked from (for source footage).
    source_media_id: str | None = None
    created_at: datetime = Field(default_factory=utcnow)
    updated_at: datetime = Field(default_factory=utcnow)
